### ISSUE TITLE
Add a parameter shift Hessian transform

### ISSUE THREAD

**Issue by josh146** [PennyLaneAI/pennylane#1752]
### Feature details

Currently, it is possible to compute the parameter-shift Hessian of a QNode by using an autodifferentiation framework, and asking for the Jacobian of the gradient (or the Jacobian of the Jacobian, for a vector-valued QNode):

```python
dev = qml.device("default.qubit", wires=2)

@qml.beta.qnode(dev, diff_method=qml.gradients.param_shift, max_diff=2)
def circuit(x):
    qml.RX(x[0], wires=0)
    qml.RY(x[1], wires=0)
    return qml.probs(wires=0)
```
```pycon
>>> x = np.array([0.1, 0.2], requires_grad=True)
>>> hess_fn = qml.jacobian(qml.jacobian(circuit))
>>> hess_fn(x)
[[[-0.48758516  0.00991692]
  [ 0.00991692 -0.48758516]]
 [[ 0.48758516 -0.00991692]
  [-0.00991692  0.48758516]]]
```

This works well if integrating the QNode into an autodifferentiation pipeline.

However, if you instead wish to compute the Hessian of a QNode in isolation, this is not ideal:

- Autodifferentiation frameworks, as they are designed for backpropagation, always evaluate the QNode, followed by the gradient, *before* computing the second derivatives. As a result, if you are *only interested in the Hessian*, there will be redundant evaluations.

- Further, if we know we only require the second derivatives, there are tricks we can use to further minimize the number of quantum evaluations. See https://arxiv.org/abs/2008.06517 and https://arxiv.org/abs/2107.12390.

- Finally, having a separate transform for computing the Hessian enables better analysis, e.g., the ability to check view the generated 'Hessian tapes', or to query the number of evaluations.

As a result, we should add a [gradient transform](https://pennylane.readthedocs.io/en/latest/code/qml_gradients.html),

```python
>>> qml.gradients.param_shift_hessian(qnode)(*args, **kwargs)
```

that returns the Hessian of a QNode _directly_, using one or more of the approaches given in https://arxiv.org/abs/2008.06517 and https://arxiv.org/abs/2107.12390.

### Implementation

- A new file should be added, `pennylane/gradients/param_shift_hessian.py`. The function should be written to be *autodiff framework agnostic* using `qml.math`, so that it will work on QNodes defined for all interfaces.

- The new function should be included in the `pennylane/gradients/__init__.py` module docstring

- Test should be added in the file `tests/gradients/test_param_shift_hessian.py`. Things to check work correctly:

  - Operators with 2-term shift rules (`qml.RX()`) should return the correct second derivatives
  - Operators with 4-term shift rules (`qml.CRX()`) should return the correct second derivatives
  - The transform should generate the minimal number of quantum evaluations.
  - The transform should be *differentiable*. That is, you should be able to do
  
    ```python
    >>> qml.grad(qml.gradients.param_shift_hessian(qnode))(weights)
    ```
    to get 3rd derivatives. Furthermore, this differentiability should work in all interfaces, including TF and Torch.

### How important would you say this feature is?

2: Somewhat important. Needed this quarter.

**Comment by dime10** [comment]
Hi @josh146, I'll tackle this one :)
